histogram.go

```diff
@@ -0,0 +1,254 @@
+	}
+	return values
+}
+
+// Durations is a set of float64 values that implements Buckets.
+type Durations []time.Duration
+
+// Implements sort.Interface
+func (v Durations) Len() int {
+	return len(v)
+}
+
+// Implements sort.Interface
+func (v Durations) Swap(i, j int) {
+	v[i], v[j] = v[j], v[i]
+}
+
+// Implements sort.Interface
+func (v Durations) Less(i, j int) bool {
+	return v[i] < v[j]
+}
+
+func (v Durations) String() string {
+	sort.Sort(v) // Always sort first
+	values := make([]string, len(v))
+	for i := range values {
+		values[i] = v[i].String()
+	}
+	return fmt.Sprintf("%v", values)
+}
+
+// AsValues implements Buckets.
+func (v Durations) AsValues() []float64 {
+	sort.Sort(v) // Always sort first
+	values := make([]float64, len(v))
+	for i := range values {
+		values[i] = float64(v[i]) / float64(time.Second)
+	}
+	return values
+}
+
+// AsDurations implements Buckets.
+func (v Durations) AsDurations() []time.Duration {
+	sort.Sort(v) // Always sort first
+	return []time.Duration(v)
+}
+
+// BucketPairs creates a set of bucket pairs from a set
+// of buckets describing the lower and upper bounds for
+// each derived bucket.
+func BucketPairs(buckets Buckets) []BucketPair {
+	var (
+		asValueBuckets    = buckets.AsValues()
+		asDurationBuckets = buckets.AsDurations()
+	)
+	var pairs []BucketPair
+	if buckets.Len() > 0 {
+		pairs = append(pairs, bucketPair{
+			-math.MaxFloat64, asValueBuckets[0],
+			time.Duration(math.MinInt64), asDurationBuckets[0]})
+
+		prevValueBucket, prevDurationBucket :=
+			asValueBuckets[0], asDurationBuckets[0]
+		for i := 1; i < buckets.Len(); i++ {
+			pairs = append(pairs, bucketPair{
+				prevValueBucket, asValueBuckets[i],
+				prevDurationBucket, asDurationBuckets[i]})
+			prevValueBucket, prevDurationBucket =
+				asValueBuckets[i], asDurationBuckets[i]
+		}
+
+		pairs = append(pairs, bucketPair{
+			prevValueBucket, math.MaxFloat64,
+			prevDurationBucket, time.Duration(math.MaxInt64)})
+	} else {
+		pairs = append(pairs, bucketPair{
+			-math.MaxFloat64, math.MaxFloat64,
+			time.Duration(math.MinInt64), time.Duration(math.MaxInt64)})
+	}
+	return pairs
+}
+
+type bucketPair struct {
+	lowerBoundValue    float64
+	upperBoundValue    float64
+	lowerBoundDuration time.Duration
+	upperBoundDuration time.Duration
+}
+
+func (p bucketPair) LowerBoundValue() float64 {
+	return p.lowerBoundValue
+}
+
+func (p bucketPair) UpperBoundValue() float64 {
+	return p.upperBoundValue
+}
+
+func (p bucketPair) LowerBoundDuration() time.Duration {
+	return p.lowerBoundDuration
+}
+
+func (p bucketPair) UpperBoundDuration() time.Duration {
+	return p.upperBoundDuration
+}
+
+// LinearValueBuckets creates a set of linear value buckets.
+func LinearValueBuckets(start, width float64, count int) Values {
+	if count <= 0 {
+		panic("count needs to be > 0")
+	}
+	buckets := make([]float64, count)
+	for i := range buckets {
+		buckets[i] = start + (float64(i) * width)
```

Comment:
The value of previous upper bounds is derived from the buckets, similar to other open source client libraries.  I believe using such small values for measurement of float64's is outside the problem domain of this implementation of histograms.

For precise time measurements using histograms using the `DurationBuckets` entirely avoids this class of problems as its using `time.Duration` which are int64 nanoseconds.